File: Data Structures & Algorithms/edit-distance/submission-1.java
class Solution {
    public int minDistance(String word1, String word2) {
        int[][] memo = new int[word1.length()][word2.length()];
        for(int[] m : memo){
            Arrays.fill(m,-1);
        }
        return dfs(0,0,memo,word1,word2);
    }

    private int dfs(int i,int j,int[][] memo,String s1,String s2){
        if(i == s1.length()) return s2.length() - j;
        if(j == s2.length()) return s1.length() - i;

        if(memo[i][j] != -1) return memo[i][j];

        if(s1.charAt(i) == s2.charAt(j)){
            memo[i][j] = dfs(i+1,j+1,memo,s1,s2);
        }else{
            memo[i][j] = 1 + Math.min(Math.min(dfs(i,j+1,memo,s1,s2),dfs(i+1,j,memo,s1,s2)),dfs(i+1,j+1,memo,s1,s2));
        }

        return memo[i][j];
    }
}

File: Data Structures & Algorithms/count-subsequences/submission-1.js
class Solution {
  /**
   * @param {string} s
   * @param {string} t
   * @return {number}
   */
  numDistinct(s, t) {
    const memo = new Map();
    const dfs = (i, j) => {
      if (j === t.length) return 1;
      if (i === s.length) return 0;
      const key = i + '-' + j;
      if (memo.has(key)) return memo.get(key);
      if (s.charAt(i) === t.charAt(j)) {
        memo.set(key, dfs(i + 1, j + 1) + dfs(i + 1, j));
      } else {
        memo.set(key, dfs(i + 1, j));
      }
      return memo.get(key);
    };
    return dfs(0, 0);
  }
}